datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -944,6 +957,101 @@ private void handleLeaderDoAssignment() {
+
+    try {
+      previousAssignmentByInstance = _adapter.getAllAssignedDatastreamTasks();
+      Map<String, Set<DatastreamTask>> assignmentByInstance = new HashMap<>(previousAssignmentByInstance);
+
+      // retrieve the datastreamGroups for validatio
+      DatastreamGroup toProcessDatastream =
+          fetchDatastreamGroups().stream().filter(dg -> datastreamGroupName.equals(dg.getName())).findFirst().orElse(null);
+
+      if (toProcessDatastream != null) {
+        AssignmentStrategy strategy = _connectors.get(toProcessDatastream.getConnectorName()).getAssignmentStrategy();
+        Connector connectorInstance = _connectors.get(toProcessDatastream.getConnectorName()).getConnector()
+            .getConnectorInstance();
+        Map<String, Optional<DatastreamGroupPartitionsMetadata>> datastreamPartitions =
+            connectorInstance.getDatastreamPartitions();
+
+        if (datastreamPartitions.containsKey(toProcessDatastream.getName())) {
+          DatastreamGroupPartitionsMetadata subscribes = connectorInstance.getDatastreamPartitions()
+              .get(toProcessDatastream.getName())
+              .orElseThrow(() ->
+                  new DatastreamTransientException("Subscribed partition is not ready yet for datastream " +
+                      toProcessDatastream.getName()));
+
+          assignmentByInstance = strategy.assignPartitions(assignmentByInstance, subscribes);
+        } else {
+          _log.warn("partitions for {} is not found, ignore the partition assignment", toProcessDatastream.getName());
+        }
+      }
+
+      for (String key : assignmentByInstance.keySet()) {
+        newAssignmentsByInstance.put(key, new ArrayList<>(assignmentByInstance.get(key)));
+      }
+      _adapter.updateAllAssignments(newAssignmentsByInstance);
+      _log.info("Partition assignment completed: datastreamGroup, assignment {} ", assignmentByInstance);
+      succeeded = true;
+    } catch (Exception ex) {
+      _log.info("Partition assignment failed, Exception: ", ex);
+    }
+    // schedule retry if failure
+    if (succeeded) {
+      _adapter.cleanupOldUnusedTasks(previousAssignmentByInstance, newAssignmentsByInstance);
+      getMaxPartitionCountInTask(newAssignmentsByInstance);
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, NUM_PARTITION_ASSIGNMENTS, 1);
+    } else if (!leaderPartitionAssignmentScheduled.get()) {
+      _log.info("Schedule retry for leader assigning tasks");
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, "handleLeaderPartitionAssignment", NUM_RETRIES, 1);
+      leaderPartitionAssignmentScheduled.set(true);
+      _executor.schedule(() -> {
+        _eventQueue.put(CoordinatorEvent.createLeaderPartitionAssignmentEvent(maybeDatastreamGroupName.get()));
+        leaderPartitionAssignmentScheduled.set(false);
+      }, _config.getRetryIntervalMs(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void getMaxPartitionCountInTask(Map<String, List<DatastreamTask>> assignments) {
```

Comment:
This methodname is misleading, it is named get but it doesn't return anything.

Reply:
Thanks, will rename the method into updateCounter